File: bindings/csharp/tests/Zag.Tests/CapabilityCheckTests.cs
using Xunit;
using Zag;

namespace Zag.Tests;

public class CapabilityCheckTests
{
    private static ProviderCapability FakeCap(
        string provider,
        bool worktree = false,
        bool sandbox = false,
        bool systemPrompt = false,
        bool addDirs = false,
        bool streamingInput = false)
    {
        return new ProviderCapability
        {
            Provider = provider,
            DefaultModel = "default",
            AvailableModels = [],
            SizeMappings = new SizeMappings(),
            Features = new Features
            {
                Interactive = new FeatureSupport { Supported = true },
                NonInteractive = new FeatureSupport { Supported = true },
                JsonOutput = new FeatureSupport { Supported = true },
                StreamJson = new FeatureSupport { Supported = true },
                AutoApprove = new FeatureSupport { Supported = true },
                Worktree = new FeatureSupport { Supported = worktree },
                Sandbox = new FeatureSupport { Supported = sandbox },
                SystemPrompt = new FeatureSupport { Supported = systemPrompt },
                AddDirs = new FeatureSupport { Supported = addDirs },
                StreamingInput = new FeatureSupport { Supported = streamingInput },
            },
        };
    }

    private static void PrimeCaps(string bin, ProviderCapability[] caps)
    {
        CapabilityCheck.ClearCapabilityCache();
        CapabilityCheck.SetAllCapabilitiesForTesting(bin, caps);
    }

    [Fact]
    public async Task Check_NoRequirements_Passes()
    {
        PrimeCaps("zag", [FakeCap("ollama")]);
        try
        {
            await CapabilityCheck.CheckAsync(
                "zag", "ollama", Array.Empty<CapabilityCheck.Requirement>());
        }
        finally
        {
            CapabilityCheck.ClearCapabilityCache();
        }
    }

    [Fact]
    public async Task Check_InactiveRequirements_Passes()
    {
        PrimeCaps("zag", [FakeCap("ollama")]);
        try
        {
            await CapabilityCheck.CheckAsync(
                "zag",
                "ollama",
                new[]
                {
                    new CapabilityCheck.Requirement("AddDir()", "add_dirs", false),
                });
        }
        finally
        {
            CapabilityCheck.ClearCapabilityCache();
        }
    }

    [Fact]
    public async Task Check_NullProvider_Skips()
    {
        // No cache primed — would raise if we tried to load.
        await CapabilityCheck.CheckAsync(
            "zag",
            null,
            new[]
            {
                new CapabilityCheck.Requirement("AddDir()", "add_dirs", true),
            });
    }

    [Fact]
    public async Task Check_MockProvider_Skips()
    {
        await CapabilityCheck.CheckAsync(
            "zag",
            "mock",
            new[]
            {
                new CapabilityCheck.Requirement("AddDir()", "add_dirs", true),
            });
    }

    [Fact]
    public async Task Check_SupportedFeature_Passes()
    {
        PrimeCaps("zag", [FakeCap("claude", streamingInput: true)]);
        try
        {
            await CapabilityCheck.CheckAsync(
                "zag",
                "claude",
                new[]
                {
                    new CapabilityCheck.Requirement(
                        "ExecStreaming()", "streaming_input", true),
                });
        }
        finally
        {
            CapabilityCheck.ClearCapabilityCache();
        }
    }

    [Fact]
    public async Task Check_UnsupportedFeature_Throws()
    {
        PrimeCaps(
            "zag",
            [
                FakeCap("claude", streamingInput: true),
                FakeCap("ollama", streamingInput: false),
            ]);
        try
        {
            var ex = await Assert.ThrowsAsync<ZagFeatureUnsupportedException>(() =>
                CapabilityCheck.CheckAsync(
                    "zag",
                    "ollama",
                    new[]
                    {
                        new CapabilityCheck.Requirement(
                            "ExecStreaming()", "streaming_input", true),
                    }));

            Assert.Equal("ExecStreaming()", ex.Method);
            Assert.Equal("streaming_input", ex.Feature);
            Assert.Equal("ollama", ex.Provider);
            Assert.Contains("claude", ex.SupportedProviders);
            Assert.DoesNotContain("ollama", ex.SupportedProviders);
            Assert.IsAssignableFrom<ZagException>(ex);
        }
        finally
        {
            CapabilityCheck.ClearCapabilityCache();
        }
    }

    [Fact]
    public async Task Check_UnsupportedWithNoSupporters_ShowsNone()
    {
        PrimeCaps("zag", [FakeCap("ollama")]);
        try
        {
            var ex = await Assert.ThrowsAsync<ZagFeatureUnsupportedException>(() =>
                CapabilityCheck.CheckAsync(
                    "zag",
                    "ollama",
                    new[]
                    {
                        new CapabilityCheck.Requirement("Sandbox()", "sandbox", true),
                    }));

            Assert.Contains("(none)", ex.Message);
        }
        finally
        {
            CapabilityCheck.ClearCapabilityCache();
        }
    }
}

public class ZagBuilderCapabilityPreflightTests
{
    private static ProviderCapability FakeCap(
        string provider,
        bool worktree = false,
        bool sandbox = false,
        bool systemPrompt = false,
        bool addDirs = false,
        bool streamingInput = false)
    {
        return new ProviderCapability
        {
            Provider = provider,
            DefaultModel = "default",
            AvailableModels = [],
            SizeMappings = new SizeMappings(),
            Features = new Features
            {
                Interactive = new FeatureSupport { Supported = true },
                NonInteractive = new FeatureSupport { Supported = true },
                JsonOutput = new FeatureSupport { Supported = true },
                StreamJson = new FeatureSupport { Supported = true },
                AutoApprove = new FeatureSupport { Supported = true },
                Worktree = new FeatureSupport { Supported = worktree },
                Sandbox = new FeatureSupport { Supported = sandbox },
                SystemPrompt = new FeatureSupport { Supported = systemPrompt },
                AddDirs = new FeatureSupport { Supported = addDirs },
                StreamingInput = new FeatureSupport { Supported = streamingInput },
            },
        };
    }

    [Fact]
    public async Task AddDir_OnOllama_Throws()
    {
        VersionCheck.SetVersionForTesting("zag", "9.9.9");
        CapabilityCheck.ClearCapabilityCache();
        CapabilityCheck.SetAllCapabilitiesForTesting("zag",
            [
                FakeCap("claude", addDirs: true),
                FakeCap("ollama", addDirs: false),
            ]);
        try
        {
            var builder = new ZagBuilder().Provider("ollama").AddDir("/extra");
            var ex = await Assert.ThrowsAsync<ZagFeatureUnsupportedException>(() =>
                builder.ExecAsync("hello"));
            Assert.Equal("AddDir()", ex.Method);
            Assert.Equal("ollama", ex.Provider);
        }
        finally
        {
            CapabilityCheck.ClearCapabilityCache();
            VersionCheck.ClearVersionCache();
        }
    }

    [Fact]
    public async Task ExecStreaming_OnGemini_Throws()
    {
        VersionCheck.SetVersionForTesting("zag", "9.9.9");
        CapabilityCheck.ClearCapabilityCache();
        CapabilityCheck.SetAllCapabilitiesForTesting("zag",
            [
                FakeCap("claude", streamingInput: true),
                FakeCap("gemini", streamingInput: false),
            ]);
        try
        {
            var builder = new ZagBuilder().Provider("gemini");
            var ex = await Assert.ThrowsAsync<ZagFeatureUnsupportedException>(() =>
                builder.ExecStreaming("hi"));
            Assert.Equal("ExecStreaming()", ex.Method);
            Assert.Equal("gemini", ex.Provider);
            Assert.Contains("claude", ex.SupportedProviders);
        }
        finally
        {
            CapabilityCheck.ClearCapabilityCache();
            VersionCheck.ClearVersionCache();
        }
    }
}

public class ZagFeatureUnsupportedExceptionTests
{
    [Fact]
    public void MessageFormat_ContainsKeyParts()
    {
        var ex = new ZagFeatureUnsupportedException(
            "ExecStreaming()",
            "streaming_input",
            "ollama",
            new[] { "claude" });

        Assert.Contains("ExecStreaming()", ex.Message);
        Assert.Contains("ollama", ex.Message);
        Assert.Contains("streaming_input", ex.Message);
        Assert.Contains("claude", ex.Message);
    }

    [Fact]
    public void EmptySupportedList_ShowsNone()
    {
        var ex = new ZagFeatureUnsupportedException(
            "Sandbox()", "sandbox", "ollama", Array.Empty<string>());
        Assert.Contains("(none)", ex.Message);
    }

    [Fact]
    public void IsZagException()
    {
        var ex = new ZagFeatureUnsupportedException(
            "Worktree()", "worktree", "x", Array.Empty<string>());
        Assert.IsAssignableFrom<ZagException>(ex);
    }
}

File: bindings/java/src/test/java/io/zag/CapabilityCheckTests.java
package io.zag;

import static org.junit.jupiter.api.Assertions.*;

import java.util.List;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;

class CapabilityCheckTests {

    private static ProviderCapability.FeatureSupport fs(boolean supported) {
        return new ProviderCapability.FeatureSupport(supported, false);
    }

    private static ProviderCapability.Features fakeFeatures(
            boolean worktree,
            boolean sandbox,
            boolean systemPrompt,
            boolean addDirs,
            boolean streamingInput) {
        return new ProviderCapability.Features(
                fs(true),                                                   // interactive
                fs(true),                                                   // nonInteractive
                fs(false),                                                  // resume
                fs(false),                                                  // resumeWithPrompt
                new ProviderCapability.SessionLogSupport(false, false, null), // sessionLogs
                fs(true),                                                   // jsonOutput
                fs(true),                                                   // streamJson
                fs(false),                                                  // jsonSchema
                fs(false),                                                  // inputFormat
                fs(streamingInput),                                         // streamingInput
                fs(worktree),                                               // worktree
                fs(sandbox),                                                // sandbox
                fs(systemPrompt),                                           // systemPrompt
                fs(true),                                                   // autoApprove
                fs(false),                                                  // review
                fs(addDirs),                                                // addDirs
                fs(false));                                                 // maxTurns
    }

    private static ProviderCapability fakeCap(
            String provider,
            boolean worktree,
            boolean sandbox,
            boolean systemPrompt,
            boolean addDirs,
            boolean streamingInput) {
        return new ProviderCapability(
                provider,
                "default",
                List.of(),
                new ProviderCapability.SizeMappings("", "", ""),
                fakeFeatures(worktree, sandbox, systemPrompt, addDirs, streamingInput));
    }

    private static ProviderCapability fakeCap(String provider) {
        return fakeCap(provider, false, false, false, false, false);
    }

    private static void primeCaps(String bin, List<ProviderCapability> caps) {
        CapabilityCheck.clearCapabilityCache();
        CapabilityCheck.setAllCapabilitiesForTesting(bin, caps);
    }

    @AfterEach
    void tearDown() {
        CapabilityCheck.clearCapabilityCache();
        VersionCheck.clearVersionCache();
    }

    // -- CapabilityCheck.check -----------------------------------------------

    @Test
    void noRequirements_returnsSilently() throws Exception {
        primeCaps("zag", List.of(fakeCap("ollama")));
        CapabilityCheck.check("zag", "ollama", List.of());
    }

    @Test
    void inactiveRequirements_returnSilently() throws Exception {
        primeCaps("zag", List.of(fakeCap("ollama")));
        CapabilityCheck.check("zag", "ollama", List.of(
                new CapabilityCheck.Requirement("addDir()", "add_dirs", false)));
    }

    @Test
    void nullProvider_skipsCheck() throws Exception {
        // No cache primed — would raise if we tried to load.
        CapabilityCheck.check("zag", null, List.of(
                new CapabilityCheck.Requirement("addDir()", "add_dirs", true)));
    }

    @Test
    void mockProvider_skipsCheck() throws Exception {
        CapabilityCheck.check("zag", "mock", List.of(
                new CapabilityCheck.Requirement("addDir()", "add_dirs", true)));
    }

    @Test
    void supportedFeature_passes() throws Exception {
        primeCaps("zag", List.of(
                fakeCap("claude", false, false, false, false, true)));
        CapabilityCheck.check("zag", "claude", List.of(
                new CapabilityCheck.Requirement("execStreaming()", "streaming_input", true)));
    }

    @Test
    void unsupportedFeature_throws() {
        primeCaps("zag", List.of(
                fakeCap("claude", false, false, false, false, true),
                fakeCap("ollama", false, false, false, false, false)));

        ZagFeatureUnsupportedException ex = assertThrows(
                ZagFeatureUnsupportedException.class,
                () -> CapabilityCheck.check("zag", "ollama", List.of(
                        new CapabilityCheck.Requirement("execStreaming()", "streaming_input", true))));

        assertEquals("execStreaming()", ex.method());
        assertEquals("streaming_input", ex.feature());
        assertEquals("ollama", ex.provider());
        assertTrue(ex.supportedProviders().contains("claude"));
        assertFalse(ex.supportedProviders().contains("ollama"));
    }

    @Test
    void unsupportedWithNoSupporters_showsNone() {
        primeCaps("zag", List.of(fakeCap("ollama")));

        ZagFeatureUnsupportedException ex = assertThrows(
                ZagFeatureUnsupportedException.class,
                () -> CapabilityCheck.check("zag", "ollama", List.of(
                        new CapabilityCheck.Requirement("sandbox()", "sandbox", true))));

        assertTrue(ex.getMessage().contains("(none)"));
    }

    // -- ZagBuilder preflight ------------------------------------------------

    @Test
    void addDirOnOllama_throwsBeforeSpawn() {
        VersionCheck.setVersionForTesting("zag", "9.9.9");
        primeCaps("zag", List.of(
                fakeCap("claude", false, false, false, true, false),
                fakeCap("ollama", false, false, false, false, false)));

        ZagFeatureUnsupportedException ex = assertThrows(
                ZagFeatureUnsupportedException.class,
                () -> new ZagBuilder()
                        .provider("ollama")
                        .addDir("/extra")
                        .exec("hello"));

        assertEquals("addDir()", ex.method());
        assertEquals("ollama", ex.provider());
        assertTrue(ex.supportedProviders().contains("claude"));
    }

    @Test
    void execStreamingOnGemini_throwsBeforeSpawn() {
        VersionCheck.setVersionForTesting("zag", "9.9.9");
        primeCaps("zag", List.of(
                fakeCap("claude", false, false, false, false, true),
                fakeCap("gemini", false, false, false, false, false)));

        ZagFeatureUnsupportedException ex = assertThrows(
                ZagFeatureUnsupportedException.class,
                () -> new ZagBuilder()
                        .provider("gemini")
                        .execStreaming("hi"));

        assertEquals("execStreaming()", ex.method());
        assertEquals("gemini", ex.provider());
        assertTrue(ex.supportedProviders().contains("claude"));
    }

    // -- ZagFeatureUnsupportedException --------------------------------------

    @Test
    void errorMessage_containsKeyParts() {
        var ex = new ZagFeatureUnsupportedException(
                "execStreaming()",
                "streaming_input",
                "ollama",
                List.of("claude"));
        assertTrue(ex.getMessage().contains("execStreaming()"));
        assertTrue(ex.getMessage().contains("ollama"));
        assertTrue(ex.getMessage().contains("streaming_input"));
        assertTrue(ex.getMessage().contains("claude"));
    }

    @Test
    void emptySupportedList_showsNone() {
        var ex = new ZagFeatureUnsupportedException(
                "sandbox()",
                "sandbox",
                "ollama",
                List.of());
        assertTrue(ex.getMessage().contains("(none)"));
    }

    @Test
    void exceptionIsZagException() {
        var ex = new ZagFeatureUnsupportedException("worktree()", "worktree", "x", List.of());
        assertTrue(ex instanceof ZagException);
    }
}
